src/main/java/io/supertokens/oauth/OAuth.java

```diff
@@ -66,39 +101,181 @@ public static OAuthAuthResponse getAuthorizationUrl(Main main, AppIdentifier app
+        // Load the JWKS from the specified URL
+        String publicOAuthProviderServiceUrl = Config.getConfig(appIdentifier.getAsPublicTenantIdentifier(), main).getOAuthProviderPublicServiceUrl();
+        String jwksUrl = publicOAuthProviderServiceUrl + HYDRA_JWKS_PATH; // Use the new constant
+
+        // Check if cached keys are available for the jwksUrl
+        Map<String, JsonObject> cachedKeys = jwksCache.get(jwksUrl);
+        if (cachedKeys == null) {
+            JsonObject jwksResponse;
+            try {
+                jwksResponse = HttpRequest.sendGETRequest(main, "", jwksUrl, null, 10000, 10000, null);
+            } catch (HttpResponseException e) {
+                throw new OAuthAPIException("jwks_error", "Could not fetch JWKS keys from hydra for token verification", 500);
+            }
+            cachedKeys = new HashMap<>();
+            JsonArray keysArray = jwksResponse.get("keys").getAsJsonArray();
+
+            // Populate the cache with keys indexed by kid
+            for (JsonElement key : keysArray) {
+                JsonObject keyObject = key.getAsJsonObject();
+                String kid = keyObject.get("kid").getAsString();
+                cachedKeys.put(kid, keyObject);
+            }
+            jwksCache.put(jwksUrl, cachedKeys); // Cache the keys with jwksUrl as the key
+        }
+
+        // Validate the JWT and extract claims using the fetched public signing keys
+        JWT.JWTPreParseInfo jwtInfo = JWT.preParseJWTInfo(token);
+        JWT.JWTInfo jwtResult = null;
+        
+        // Check if the key for the given kid exists in the cache
+        JsonObject keyObject = cachedKeys.get(jwtInfo.kid);
+        if (keyObject != null) {
+            jwtResult = JWT.verifyJWTAndGetPayload(jwtInfo, keyObject.get("n").getAsString(), keyObject.get("e").getAsString());
+        }
+
+        if (jwtResult == null) {
+            // If no matching key found and retry count is not exceeded, refetch the keys
+            if (retryCount < MAX_RETRIES) {
+                jwksCache.remove(jwksUrl); // Invalidate cache
+                return reSignToken(appIdentifier, main, token, iss, stt, useDynamicSigningKey, retryCount + 1); // Retry with incremented count
             } else {
-                throw new RuntimeException("Unexpected answer from Oauth Provider");
+                throw new OAuthAPIException("invalid_token", "The token is invalid or has expired.", 400);
             }
-            if(responseHeaders.containsKey(COOKIES_HEADER_NAME)){
-                cookies = responseHeaders.get(COOKIES_HEADER_NAME);
+        }
+
+        JsonObject payload = jwtResult.payload;
+        // move keys in ext to root
+        if (payload.has("ext")) {
```

Comment:
I don't think the id_token has `ext` (but please double check this) so if it's there, we may want to leave it alone.

Reply:
Done